### PR TITLE
Propagate variant parse failures

### PR DESCRIPTION
**Pull request by nh13**
Previously iteration would silently stop.

https://github.com/brentp/cyvcf2/issues/161

### CODE REVIEW

File: cyvcf2/cyvcf2.pyx [hunk 1]
```diff
@@ -570,11 +570,11 @@ cdef class VCF(HTSFile):
         with nogil:
             b = bcf_init()
             ret = bcf_read(self.hts, self.hdr, b)
-        if ret >= 0:
+        if ret >= 0 || b->error == BCF_ERR_CTG_UNDEF:
             return newVariant(b, self)
```
[the remark this answers is not in the document]
Comment (brentp) [reply cyvcf2/cyvcf2.pyx:574]: shouldn't it be `or` instead of `||`?

File: cyvcf2/cyvcf2.pyx [hunk 2]
```diff
@@ -570,11 +570,11 @@ cdef class VCF(HTSFile):
...
         else:
             bcf_destroy(b)
-        raise StopIteration
```
Comment (brentp) [inline cyvcf2/cyvcf2.pyx:577]: this won't work. need to differentiate between end of file and other modes.

File: cyvcf2/cyvcf2.pyx [hunk 3]
```diff
@@ -570,11 +570,11 @@ cdef class VCF(HTSFile):
         with nogil:
             b = bcf_init()
             ret = bcf_read(self.hts, self.hdr, b)
-        if ret >= 0:
+        if ret >= 0 || b->error == BCF_ERR_CTG_UNDEF:
```
Comment (nh13) [inline cyvcf2/cyvcf2.pyx:573]:
```suggestion
        if ret >= 0 or b->error == BCF_ERR_CTG_UNDEF:
```

File: cyvcf2/cyvcf2.pyx (hunk 1 above)
[the remark this answers is not in the document]
Comment (nh13) [reply cyvcf2/cyvcf2.pyx:574]: What language am I writing in again?  It's been a long day, than-you!

File: cyvcf2/cyvcf2.pyx [hunk 4]
```diff
@@ -570,11 +570,11 @@ cdef class VCF(HTSFile):
         with nogil:
             b = bcf_init()
             ret = bcf_read(self.hts, self.hdr, b)
-        if ret >= 0:
+        if ret >= 0 or b->error == BCF_ERR_CTG_UNDEF:
             return newVariant(b, self)
         else:
             bcf_destroy(b)
-        raise StopIteration
+            raise Exception("error parsing variant with `htslib::bcf_read` error-code: %d" % (b->err))
```
Comment (nh13) [inline cyvcf2/cyvcf2.pyx:577]:
```suggestion
        if  ret == -1:  # end-of-file
            raise StopIteration
        else:  
            raise Exception("error parsing variant with `htslib::bcf_read` error-code: %d" % (b->err))

```

File: cyvcf2/cyvcf2.pyx [hunk 5]
```diff
@@ -570,11 +570,15 @@ cdef class VCF(HTSFile):
         with nogil:
             b = bcf_init()
             ret = bcf_read(self.hts, self.hdr, b)
-        if ret >= 0:
+        if ret >= 0 or b->error == BCF_ERR_CTG_UNDEF:
             return newVariant(b, self)
         else:
             bcf_destroy(b)
-        raise StopIteration
+        if  ret == -1:  # end-of-file
```
Comment (nh13) [inline cyvcf2/cyvcf2.pyx:577]: See: https://github.com/samtools/htslib/blob/4162046b28a7d9d8a104ce28086d9467cc05c212/htslib/vcf.h#L379

**Comment by nh13** [comment]
@brentp thanks for the quick review, I've fixed up the mistakes you caught.

File: cyvcf2/cyvcf2.pyx [hunk 6]
```diff
@@ -570,11 +570,15 @@ cdef class VCF(HTSFile):
...
-        if ret >= 0:
+        if ret >= 0 or b.error == BCF_ERR_CTG_UNDEF:
             return newVariant(b, self)
         else:
             bcf_destroy(b)
-        raise StopIteration
+        if  ret == -1:  # end-of-file
+            raise StopIteration
+        else:  
+            raise Exception("error parsing variant with `htslib::bcf_read` error-code: %d" % (b.ferr))
```
Comment (nh13) [inline cyvcf2/cyvcf2.pyx:580]:
```suggestion
            raise Exception("error parsing variant with `htslib::bcf_read` error-code: %d" % (b.err))
```

**Comment by nh13** [comment]
Editing this on mobile is "fun".  Sorry @brentp for the spam.

File: cyvcf2/cyvcf2.pyx [hunk 7]
```diff
@@ -570,11 +570,15 @@ cdef class VCF(HTSFile):
...
-        if ret >= 0:
+        if ret >= 0 or b.error == BCF_ERR_CTG_UNDEF:
             return newVariant(b, self)
         else:
             bcf_destroy(b)
-        raise StopIteration
+        if  ret == -1:  # end-of-file
+            raise StopIteration
+        else:  
+            raise Exception("error parsing variant with `htslib::bcf_read` error-code: %d" % (b.err))
```
Comment (nh13) [inline cyvcf2/cyvcf2.pyx:580]:
```suggestion
            raise Exception("error parsing variant with `htslib::bcf_read` error-code: %d" % (b.error))
```

File: cyvcf2/cyvcf2.pyx [hunk 8]
```diff
@@ -570,11 +570,15 @@ cdef class VCF(HTSFile):
         with nogil:
             b = bcf_init()
             ret = bcf_read(self.hts, self.hdr, b)
-        if ret >= 0:
+        if ret >= 0 or b.error == BCF_ERR_CTG_UNDEF:
```
Comment (nh13) [inline cyvcf2/cyvcf2.pyx:573]:
```suggestion
        if ret >= 0 or b.errcode == BCF_ERR_CTG_UNDEF:
```

File: cyvcf2/cyvcf2.pyx [hunk 9]
```diff
@@ -570,11 +570,15 @@ cdef class VCF(HTSFile):
...
             return newVariant(b, self)
         else:
             bcf_destroy(b)
-        raise StopIteration
+        if  ret == -1:  # end-of-file
+            raise StopIteration
+        else:  
+            raise Exception("error parsing variant with `htslib::bcf_read` error-code: %d" % (b.error))
```
Comment (nh13) [inline cyvcf2/cyvcf2.pyx:580]:
```suggestion
            raise Exception("error parsing variant with `htslib::bcf_read` error-code: %d" % (b.errcode))
```

**Comment by nh13** [comment]
@brentp all checks pass!

**Comment by brentp** [comment]
thanks.

**Pull request merged**